src/daft-core/src/array/ops/concat_agg.rs

```diff
@@ -146,6 +149,43 @@ impl DaftConcatAggable for ListArray {
     }
 }
 
+impl DaftConcatAggable for DataArray<Utf8Type> {
+    type Output = DaftResult<Self>;
+
+    fn concat(&self) -> Self::Output {
+        let mut concat_result = String::new();
+
+        for idx in 0..self.len() {
+            if self.get(idx).is_some() {
+                let x = self.get(idx).unwrap().to_string();
+                concat_result.push_str(&x);
+            }
+        }
+        let result_box = Box::new(arrow2::array::Utf8Array::<i64>::from_slice([concat_result]));
```

Comment:
So, simply creating a String and calling `push_str` can be inefficient due to potentially repeated heap allocations.

Instead, we only want one heap allocation. The easy way would be to do `String::with_capacity`, but, if you take a look at arrow2::array::Utf8Array, you should notice that it stores values as a `Buffer<u8>`. Therefore, all you really need to do to is to clone the buffer and modify the offsets to reflect the whole buffer as a single string, e.g. `let new_offsets = OffsetsBuffer::<i64>::try_from(vec![0, *self.as_arrow.offsets().last()])?;`

Note: make sure to account for cases where the whole array may be null.

Reply:
@colin-ho In case of all nulls, the datatype was not Utf8.  In that case, what can we do? can we return an empty string or `[None]`?

Reply:
You can make a UTF8 array with all nulls via casting: `df.select(daft.col('data').cast(daft.DataType.String()))`

Reply:
In which case, the output of `agg_concat` should be `[None]`

Reply:
ok.  Will look into it.  Thanks

Reply:
@colin-ho I made the changes as per the comments for both concat and grouped concat on string and also with null list.  Please let me know if any other changes were needed.